File: ruby-elasticemail-examples/examples/domains.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

##
# Domain Management
#
# Adds a sending domain and prints its DNS verification state.
#
# Usage: ruby examples/domains.rb

require_relative "ee"

domains_api = ElasticEmail::DomainsApi.new

domain = SENDING_DOMAIN

# 1. Add the domain
begin
  domains_api.domains_post(ElasticEmail::DomainPayload.new(domain: domain))
  puts "Domain \"#{domain}\" added."
rescue ElasticEmail::ApiError => e
  if already_exists?(e)
    puts "Domain \"#{domain}\" already exists."
  else
    api_error("add domain", e)
  end
end

# 2. Show DNS verification state. Add the records shown in the dashboard, then re-run.
begin
  data = domains_api.domains_by_domain_get(domain)
  puts "\nVerification status:"
  puts "  SPF:       #{data.spf ? 'ok' : 'missing'}"
  puts "  DKIM:      #{data.dkim ? 'ok' : 'missing'}"
  puts "  MX:        #{data.mx ? 'ok' : 'missing'}"
  puts "  DMARC:     #{data.dmarc ? 'ok' : 'missing'}"
  puts "  Tracking:  #{data.tracking_status || 'n/a'}"
  puts "  Default:   #{data.default_domain ? 'yes' : 'no'}"
  if data.dkim_record
    puts "\nDKIM record to publish: #{data.dkim_record.to_hash.to_json}"
  end
rescue ElasticEmail::ApiError => e
  api_error("get domain", e)
end

# 3. List all domains
begin
  domains = domains_api.domains_get
  puts "\nDomains on the account (#{domains.length}):"
  domains.each do |d|
    puts " - #{d.domain} spf=#{d.spf} dkim=#{d.dkim} default=#{d.default_domain}"
  end
rescue ElasticEmail::ApiError => e
  api_error("list domains", e)
end

# 4. Optional: verify tracking (CNAME must point at the Elastic Email tracking host)
# domains_api.domains_by_domain_verification_put(domain, "Http")

# 5. Optional: set the default sender for the account
# domains_api.domains_by_email_default_patch("hello@#{domain}")

puts "\nDone."

File: elixir-elasticemail-examples/examples/domains.exs
alias ElasticEmail.Env

domain = Env.sending_domain()
domain_path = ElasticEmail.path_segment(domain)

fail = fn step, error ->
  IO.puts(:stderr, "Error (#{step}): #{ElasticEmail.format_error(error)}")
  System.halt(1)
end

flag = fn value -> if value, do: "ok", else: "missing" end

# 1. Add the domain
case ElasticEmail.post("/domains", %{"Domain" => domain}) do
  {:ok, _} ->
    IO.puts(~s(Domain "#{domain}" added.))

  {:error, 400, body} = error ->
    if inspect(body) =~ ~r/exist|already/i,
      do: IO.puts(~s(Domain "#{domain}" already exists.)),
      else: fail.("add domain", error)

  error ->
    fail.("add domain", error)
end

# 2. Show DNS verification state. Add the records shown in the dashboard, then re-run.
case ElasticEmail.get("/domains/#{domain_path}") do
  {:ok, d} ->
    IO.puts("\nVerification status:")
    IO.puts("  SPF:       #{flag.(d["Spf"])}")
    IO.puts("  DKIM:      #{flag.(d["Dkim"])}")
    IO.puts("  MX:        #{flag.(d["MX"])}")
    IO.puts("  DMARC:     #{flag.(d["DMARC"])}")
    IO.puts("  Tracking:  #{d["TrackingStatus"] || "n/a"}")
    IO.puts("  Default:   #{if d["DefaultDomain"], do: "yes", else: "no"}")

    if d["DKIMRecord"] do
      IO.puts("\nDKIM record to publish: #{Jason.encode!(d["DKIMRecord"])}")
    end

  error ->
    fail.("get domain", error)
end

# 3. List all domains
case ElasticEmail.get("/domains") do
  {:ok, domains} ->
    IO.puts("\nDomains on the account (#{length(domains)}):")

    Enum.each(domains, fn d ->
      IO.puts(" - #{d["Domain"]} spf=#{d["Spf"]} dkim=#{d["Dkim"]} default=#{d["DefaultDomain"]}")
    end)

  error ->
    fail.("list domains", error)
end

# 4. Optional: verify tracking (CNAME must point at the Elastic Email tracking host)
# ElasticEmail.put("/domains/#{domain_path}/verification", "Http")

# 5. Optional: set the default sender for the account
# ElasticEmail.patch("/domains/#{ElasticEmail.path_segment("hello@" <> domain)}/default", %{})

IO.puts("\nDone.")
